core/src/main/scala/io/dantb/contentless/ContentModel.scala

```diff
@@ -85,20 +95,60 @@ object Validation:
-
-  // equality defers to underlying regex, to allow custom regex subtypes for convenience without breaking equality
-  sealed abstract case class Regexp(regexp: String) extends Validation:
-    override def equals(obj: Any): Boolean = obj.isInstanceOf[Regexp] && obj.asInstanceOf[Regexp].regexp == regexp
-    override def hashCode(): Int           = regexp.hashCode
-  object ValidUrl extends Regexp(Regexp.validUrlRegexp)
-  object Regexp:
-    val validUrlRegexp = "^(ftp|http|https):\\/\\/(\\w+:{0,1}\\w*@)?(\\S+)(:[0-9]+)?(\\/|\\/([\\w#!:.?+=&%@!\\-\\/]))?$"
 
-    def apply(str: String): Regexp = new Regexp(str) {}
+  final case class RichTextMarks(allowedValues: Set[Mark]) extends Validation:
+    def message: Option[String] =
+      val size = allowedValues.size
+      size match
+        case 0 => None
+        case 1 => s"Only ${allowedValues.head.asString} marks are allowed".some
+        case other =>
+          val (allButOne, last) = allowedValues.map(_.asString).splitAt(size - 1)
+          s"Only ${allButOne.mkString(", ")} and ${last.head} marks are allowed".some
+
+  final case class RichTextNodeTypes(allowedValues: Set[RichTextNodeType]) extends Validation:
+    def message: Option[String] =
+      val size = allowedValues.size
+      size match
+        case 0 => None
+        case 1 => s"Only ${allowedValues.head.asString} nodes are allowed".some
+        case other =>
+          val (allButOne, last) = allowedValues.map(_.asString).splitAt(size - 1)
+          s"Only ${allButOne.map(_.replaceAll("-", " ")).mkString(", ")} and ${last.head} nodes are allowed".some
+
+  final case class Size(min: Option[Int], max: Option[Int], message: Option[String]) extends Validation
+  final case class DateRange(min: Option[ZonedDateTime], max: Option[ZonedDateTime]) extends Validation
+  case object Unique                                                                 extends Validation
+
+  enum Regexp(val underlying: Regex) extends Validation:
+    case Url
+        extends Regexp("^(ftp|http|https):\\/\\/(\\w+:{0,1}\\w*@)?(\\S+)(:[0-9]+)?(\\/|\\/([\\w#!:.?+=&%@!\\-\\/]))?$".r)
+    case Email                  extends Regexp("^\\w[\\w.-]*@([\\w-]+\\.)+[\\w-]+$".r)
+    case Slug                   extends Regexp("^[a-z0-9]+(?:-[a-z0-9]+)*$".r)
+    case DateUS                 extends Regexp("^(0?[1-9]|1[012])[- /.](0?[1-9]|[12][0-9]|3[01])[- /.](19|20)?\\d\\d$".r)
+    case DateEurope             extends Regexp("^(0?[1-9]|[12][0-9]|3[01])[- /.](0?[1-9]|1[012])[- /.](19|20)?\\d\\d$".r)
+    case Time12H                extends Regexp("^(0?[1-9]|1[012]):[0-5][0-9](:[0-5][0-9])?\\s*[aApP][mM]$".r)
+    case Time24H                extends Regexp("^(0?[0-9]|1[0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?$".r)
+    case PhoneNumUS             extends Regexp("^\\d[ -.]?\\(?\\d\\d\\d\\)?[ -.]?\\d\\d\\d[ -.]?\\d\\d\\d\\d$".r)
+    case ZipCodeUS              extends Regexp("^\\d{5}$|^\\d{5}-\\d{4}$}".r)
```

Comment:
Are these regexes defined by contentful?

Reply:
Yeah I copied them directly. If they add more it should still be compatible and we could introduce later